### ISSUE TITLE
Backtest Friendly Dataframe Lookback in Populate_*

### ISSUE THREAD

**Issue by BobbyBobbyTheBobBob** [freqtrade/freqtrade#5155]
## Describe your environment
  * Freqtrade Version:  Docker  
## Your question

I started to title this: “Explain Backtesting to me like I’m 5” though decided on something more useful… but that’s where I am.

As I read (and read and read) the docs and assorted posts, I understand there is no backtest friendly way for a custom indicator to lookback at previous rows of the dataframe.

(If I’m incorrect, tell me now because…)

That leads to to two questions:

How do MAs and others do it? (If you can explain it to a 5 year old) Obviously they work differently, I’m curious how.
 
Will this workaround be backtest friendly:  
1)  I [store an array (list) in the strategy class](https://www.freqtrade.io/en/stable/strategy-advanced/#storing-information) that will hold the previous n candles of values.  
2)  Using [bot_loop_start() ](https://www.freqtrade.io/en/stable/bot-basics/#bot-execution-logic)I get the current value and put it in the list. 
3)  n candles later, I can read it with my strategy instead of using the dataframe.

If you can’t explain the WHY, that’s fine, but can someone at least confirm my method would be backtest friendly. 

I'm asking vs just trying it because: First, I'm -really- new and that will take me like 3 hours to figure out and Second, it seems like a ton of people are trying to lookback and if my workaround is valid maybe this question will save others time in the future. 

Thanks

**Comment by xmatthias** [comment]
Best look for a pandas tutorial concerning time-series. These usually cover these topics.

You can very easily look back a few rows, either with `.shift(lookback)` - which will look back exactly that many rows - or `.rolling(lookback).<somefunction>()` - which will for example calculate an SMA (`.rolling(20).mean()` would be SMA20).

You might wanna look at the [bot basics](https://www.freqtrade.io/en/stable/bot-basics/#backtesting-hyperopt-execution-logic) documentation - which explains the difference in execution logic between backtesting and trade mode.

to put it simply - your approach will not work for backtesting - as backtesting calls each `populate_*()` method once per pair (with the full dataframe)- so your calculations should be vectorized (again, use your favorite search-engine with vectorized + pandas to understand this topic).

**Comment by BobbyBobbyTheBobBob** [comment]
Many thanks I've seen you say to vectorize but didn't know what it meant and didn't think to search it + pandas.

I'll beat on it some more.

**Comment by BobbyBobbyTheBobBob** [comment]
Well THAT's embarrassing... When you link something in your question and that exact link tells you that you're wrong. 

Sorry, xmatthias I clearly read that link but missed the line that told me I was wrong. Clown nose on. My bad.

**Comment by xmatthias** [comment]
I'll assume this as answered by the above.

Feel free to comment below / reopen if that's not the case.
